compatibility.bs

```diff
@@ -832,6 +847,129 @@ supported on all {{Window}} objects and <code><{body}></code> elements as attrib
+1. Set [=heuristic grants map=][|key|] to |expirationTime|.
+
+</div>
+
+<h3 id="sa-heuristics-access">Cookie Access Monkey Patch</h3>
+
+Issue: TODO: add monkey patch to network fetch cookie access that reads from [=heuristic grants map=].
+
+<h3 id="sa-heuristics-popup">Popup Heuristic</h3>
+
+Append the following steps to the [[!HTML]]'s <a spec="html">activation notification</a> algorithm:
+
+<div algorithm="user-activation-patch">
+
+1. Run [=detect a popup heuristic=] given <var ignore>document</var>.
+
+</div>
+
+<div algorithm>
+
+To <dfn>detect a popup heuristic</dfn> given a [=Document=] |document|,
+perform the following steps:
+
+1. Let |browsingContext| be |document|'s [=Document/browsing context=].
+1. If |browsingContext|'s [=browsing context/is popup=] is false, then return.
+1. If |browsingContext|'s [=browsing context/opener browsing context=] is null, then return.
+1. If |browsingContext|'s [=browsing context/opener origin at creation=] is null, then return.
+1. Let |firstPartySite| be the result of running [=obtain a site=] given the |browsingContext|'s
+    [=browsing context/opener origin at creation=].
+1. Let |navigable| be |document|'s [=node navigable=].
+1. If |navigable| is null, then return.
+1. Let |topDocument| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=navigable/active document=].
+1. Let |origin| be |topDocument|'s [=Document/origin=].
+1. If |origin| is an [=opaque origin=] then return.
+1. Let |site| be the result of running [=obtain a site=] given |origin|.
+1. Let |currentWallTime| be |topDocument|'s [=relevant settings object=]'s
+    [=environment settings object/current wall time=].
+1. [=Grant access for heuristics=] given |site|, |firstPartySite|, |currentWallTime|, and
+    [=popup heuristic grant duration=].
+
+Issue(amaliev/3pcd-exemption-heuristics#3): TODO: Consider whether to check for third-party iframe initiators.
+
+</div>
+
+<h3 id="sa-heuristics-redirect">Redirect Heuristic</h3>
+
+Insert the following steps in [[!HTML]]'s <a spec="html">load a document</a> algorithm, before the
+final step which returns null.
+
+<div algorithm="loading-a-document-patch">
+
+1. Run [=detect a redirect heuristic=] given <var ignore>navigable</var>.
+
+</div>
+
+<div algorithm>
+
+To <dfn>detect a redirect heuristic</dfn> given a [=/navigable=] |navigable|, perform the following
+steps:
+
+1. Let |topDocument| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=navigable/active document=].
+1. Let |firstPartyOrigin| be |topDocument|'s [=Document/origin=].
+1. If |firstPartyOrigin| is an [=opaque origin=] then return.
+1. Let |firstPartySite| be the result of running [=obtain a site=] given |firstPartyOrigin|.
+1. Let |bounceTrackingRecord| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=top-level traversable/bounce tracking record=].
+1. [=list/For each=] |bounceUrl| in |bounceTrackingRecord|'s [=bounce tracking record/bounce set=]:
```

Comment:
Does chrome's implementation follow this? If you use the bounce record, you are relying on the user going through the entire "auth" page in <10 seconds, since the records are purged after 10 seconds on one page.

Reply:
Ah, maybe I misunderstood. The extended navigation here is just the B->A to end it? and that last TODO is looking for the A->B in the same navigatable's history to complete the A->B->A?

Reply:
Thanks for the catch. I've updated the nav tracking spec to add the initial host to the user activation set in both cases.